File: src/web/Backend/src/utils/searchKind.js
const ART_ID_PREFIX = "graf";
const ART_ID_TOTAL_LEN = 64;

const isHex64 = (s) => /^[0-9a-fA-F]{64}$/.test(s || "");

function guessKind(raw) {
  const q = (raw || "").trim();
  if (!q) return "unknown";

  const lower = q.toLowerCase();
  if (lower.startsWith(ART_ID_PREFIX)) return "art_id";
  if (lower.startsWith("tsar") && q.length >= 20) return "address";
  if (/^\d{1,7}$/.test(q)) return "block_height";
  
  if (isHex64(q)) {
    return "hash64";
  }
  
  return "unknown";
}

module.exports = { guessKind, isHex64, ART_ID_PREFIX, ART_ID_TOTAL_LEN };

File: src/web/Backend/src/utils/search_kind.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2025 Tsar Studio
# Part of TsarChain — see LICENSE

import re

ART_ID_PREFIX = "graf"

_HEX64_REGEX = re.compile(r"^[0-9a-fA-F]{64}$")
_BLOCK_HEIGHT_REGEX = re.compile(r"^\d{1,7}$")


def is_hex64(s: str | None) -> bool:
    if not s or type(s) is not str:
        return False
    return bool(_HEX64_REGEX.match(s))


def guess_kind(raw: str | None) -> str:
    if not raw or type(raw) is not str:
        return "unknown"
    q = raw.strip()
    if not q:
        return "unknown"

    lower = q.lower()
    if lower.startswith(ART_ID_PREFIX):
        return "art_id"
    if lower.startswith("tsar") and len(q) >= 20:
        return "address"
    if _BLOCK_HEIGHT_REGEX.match(q):
        return "block_height"
    if _HEX64_REGEX.match(q):
        return "hash64"

    return "unknown"
